File: 卡码/代码随想录/Go/回溯/02-分割/LeetCode_0131.go
//131. 分割回文串
package main

func isPalindrome(s string,start,end int) bool{
	for start<end {
		if s[start]!=s[end] {
			return false
		}
		start++
		end--
	}
	return true
}

func partition(s string) [][]string {
	var path=make([]string,0)
	var res=make([][]string,0)
	var backtracking func(s string,index int)
	backtracking=func(s string,index int){
		if index>=len(s) {
			tmp:=make([]string,len(path))
			copy(tmp,path)
			res=append(res,tmp)
			return
		}
		for i:=index;i<len(s);i++ {
			if isPalindrome(s,index,i) {
				path=append(path,s[index:i+1])
			}else{
				continue
			}
			backtracking(s,i+1)
			path=path[:len(path)-1]
		}
	}
	backtracking(s,0)
	return res
}

File: 卡码/代码随想录/Python/回溯/02-分割/LeetCode_0131.py
#131. 分割回文串
from typing import *

class Solution:
  def __init__(self):
    self.path:List[str]=[]
    self.res:List[List[str]]=[]
  def isPalindrome(self,s:str,start:int,end:int)->bool:
    while start<end:
      if s[start]!=s[end]:
        return False
      start+=1
      end-=1
    return True
  def backtracking(self,s:str,index:int):
    if index>=len(s):
      self.res.append(self.path[:])
      return
    for i in range(index,len(s)):
      if self.isPalindrome(s,index,i):
        self.path.append(s[index:i+1])
      else:
        continue
      self.backtracking(s,i+1)
      self.path.pop()
  def partition(self, s: str) -> List[List[str]]:
    self.backtracking(s,0)
    return self.res